rpc/examples/echo/frontend/src/index.ts

```diff
@@ -57,101 +60,90 @@ async function getClients() {
+		const webrtcClient = createPromiseClient(EchoService, webRTCConn.transport);
+		await renderResponses(webrtcClient, "wrtc");
+	} catch (err) {
+		console.error("error trying WebRTC", err);
+	}
 
-	const directTransport = await dialDirect(thisHost, opts);
-	const directClient = new EchoServiceClient(thisHost, { transport: directTransport });
-	await renderResponses(directClient, "direct");
+	try {
+		const directTransport = await dialDirect(thisHost, opts);
+		const directClient = createPromiseClient(EchoService, directTransport);
+		await renderResponses(directClient, "direct");
+	} catch (err) {
+		console.error("error trying direct", err);
+	}
 }
+
 getClients().catch(e => {
 	console.error("error getting clients", e);
 });
 
-async function renderResponses(client: EchoServiceClient, method: string) {
+async function renderResponses(client: PromiseClient<typeof EchoService>, method: string) {
 	const echoRequest = new EchoRequest();
-	echoRequest.setMessage("hello");
-
-	let pResolve: (value: any) => void;
-	let pReject: (reason?: any) => void;
-	let done = new Promise<any>((resolve, reject) => {
-		pResolve = resolve;
-		pReject = reject;
-	});
-	client.echo(echoRequest, (err: ServiceError, resp: EchoResponse) => {
-		if (err) {
-			console.error(err);
-			pReject(err);
-			return
-		}
-		createElemForResponse(resp.getMessage(), method, "unary");
-		pResolve(resp);
-	});
-	await done;
+	echoRequest.message = "hello";
+
+	const response = await client.echo(echoRequest);
+	createElemForResponse(response.message, method, "unary");
 
 	const echoMultipleRequest = new EchoMultipleRequest();
-	echoMultipleRequest.setMessage("hello?");
+	echoMultipleRequest.message = "hello?";
 
-	done = new Promise<any>((resolve, reject) => {
-		pResolve = resolve;
-		pReject = reject;
-	});
 	const multiStream = client.echoMultiple(echoMultipleRequest);
-	multiStream.on("data", (message: EchoMultipleResponse) => {
-		createElemForResponse(message.getMessage(), method, "multi");
-	});
-	multiStream.on("end", ({ code, details }: { code: number, details: string, metadata: grpc.Metadata }) => {
-		if (code !== 0) {
-			console.log(code);
-			console.log(details);
-			pReject(code);
-			return;
+	try {
+		for await (const response of multiStream) {
+			createElemForResponse(response.message, method, "multi");
 		}
-		pResolve(undefined);
-	});
-	await done;
-
-	const bidiStream = client.echoBiDi();
+	} catch (err) {
+		if (err instanceof ConnectError) {
+			console.log(err.code);
+			console.log(err.details);
+		}
+		throw err;
+	}
 
-	let echoBiDiRequest = new EchoBiDiRequest();
-	echoBiDiRequest.setMessage("one");
+	if (method === "direct") {
+		// grpc-web cannot do bidi correctly. Previous versions
+		// of this code just stalled after the first message was received.
+		return;
+	}
 
-	done = new Promise<any>((resolve, reject) => {
-		pResolve = resolve;
-		pReject = reject;
-	});
+	const clientStream = createWritableIterable<PartialMessage<EchoBiDiRequest>>();
```

Comment:
Look around here for how to use connect-es. IMO, a lot more ergonomic. Even the bidi case isn't that nasty